docs/adrs/0002-uds-core-functional-layers.md

```diff
@@ -0,0 +1,73 @@
+# ADR: UDS Core Functional Layers
+
+Date: 2024-07-17
+
+## Status
+
+Accepted
+
+## Context
+
+Historically, UDS Core intentionally only published a single official Zarf Package to ensure a standard baseline across all UDS environments. We learned from the complex buffet menu-style of Big Bang that it is better to form opinions and test them than to allow unlimited configuration possibilities. We have also found the monorepo structure of UDS Core to be be much simpler for developing UDS Core. However, through user feedback, we have seen cases where more optionality is warranted, but having to modify UDS CLI to accommodate disabling portions of UDS Core is not ideal. We also have a need for an edge-compatible version of UDS Core that is more lightweight and does not include all the components of the full UDS Core package.
+
+Today, we publish both the official UDS Core package and a UDS Slim Dev package, in addition to dev and demo UDS Bundles. The UDS Slim Dev package is a stripped-down version of UDS Core that includes limited services meant for testing UDS Packages. We have also learned that some teams outside of UDS need similar functionality, but slightly different parts of UDS Core, such as keycloak + authsvc.
+
+## Decision
+
+We will replace the existing 2-package (one dev and one prod) structure with a layered set of packages within the existing monorepo. This will allow users to choose the components they need for their specific use case, while maintaining a single source of truth for all UDS Core components. We will not break these out by applications as Big Bang did, but instead by layers of functionality for related/integrated components. This allows more flexibility for which functionality is provided by UDS Core and which is provided by other external packages.
+
+The functional layers will be:
+
+1. UDS Core Base (Istio + Pepr w/UDS Operator & UDS Policies)
+2. UDS Core Identity and Authorization (Keycloak + AuthService)
+3. UDS Core Metrics (Metrics Server)
+4. UDS Core Monitoring (Prometheus + Grafana)
+5. UDS Core Logging (Promtail + Loki)
+6. UDS Core Runtime Security (NeuVector)
+7. UDS Core Backup and Restore (Velero)
+8. UDS Core (combination of all layers)
```

Comment:
It may be worth considering a future where UDS Core does not have all the layers by default, or has all layers but some are optional. Maybe could be saved for a future ADR but two cases I am thinking of:
1. Metrics-server: Most distros either provide metrics-server out of the box (rke2, k3s/k3d) or provide an easy/standard addon for it (EKS). To deploy core today someone would have to explicitly disable RKE2/k3d's metrics-server ([ex in uds-k3d](https://github.com/defenseunicorns/uds-k3d/blob/main/zarf.yaml#L48)). This might be a good case for making it in core but optional.
2. Kiali/Tempo: Both have come up in the past and could be useful but there are varied opinions on them. If neither provides any additional compliance value then they may not necessary as defaults in core, but could still be valuable as a `service-mesh-addons` type package, optional in core.

Reply:
> It may be worth considering a future where UDS Core does not have all the layers by default, or has all layers but some are optional. Maybe could be saved for a future ADR but two cases I am thinking of:
> 
>     1. Metrics-server: Most distros either provide metrics-server out of the box (rke2, k3s/k3d) or provide an easy/standard addon for it (EKS). To deploy core today someone would have to explicitly disable RKE2/k3d's metrics-server ([ex in uds-k3d](https://github.com/defenseunicorns/uds-k3d/blob/main/zarf.yaml#L48)). This might be a good case for making it in core but optional.

Are there any good patterns we could leverage to conditionally deploy metrics server (i.e. if the capability exists on the cluster, skip this component install). That may help reduce optional levers for bundle creators.

Reply:
Yeah absolutely - with zarf this isn't really possible (conditional components based on a cluster check), but with uds-cli we can build it to make it work for us. As the other active PR indicates, to start we'll likely start with a manual optional, but move to building out an automated method of detection in CLI.

Reply:
Add Azure AKS to that list.  We're having to work around the native metrics server.  I'm sure we can come up with some consistent patterns to skip for pre-existing services.

Reply:
Did not make any specific changes to the ADR based on this thread - I think it still makes sense for Core itself to contain all the layers, even if some are optional. The specific convo on metrics-server here will be addressed separately as a part of https://github.com/defenseunicorns/uds-core/pull/611 and the corresponding issue.